src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -3292,5 +3293,88 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
 
             return -1;
         }
+
+        public static int Count<T>(ref T current, T value, int length) where T : IEquatable<T>?
+        {
+            int count = 0;
+
+            ref T end = ref Unsafe.Add(ref current, length);
+            if (value is not null)
+            {
+                while (Unsafe.IsAddressLessThan(ref current, ref end))
+                {
+                    if (value.Equals(current))
+                    {
+                        count++;
+                    }
+
+                    current = ref Unsafe.Add(ref current, 1);
+                }
+            }
+            else
+            {
+                while (Unsafe.IsAddressLessThan(ref current, ref end))
+                {
+                    if (current is null)
+                    {
+                        count++;
+                    }
+
+                    current = ref Unsafe.Add(ref current, 1);
+                }
+            }
+
+            return count;
+        }
+
+        public static int CountValueType<T>(ref T current, T value, int length) where T : struct, IEquatable<T>?
+        {
+            int count = 0;
+
+            ref T end = ref Unsafe.Add(ref current, length);
+            if (Vector128.IsHardwareAccelerated && length >= Vector128<T>.Count)
+            {
+                if (Vector256.IsHardwareAccelerated && length >= Vector256<T>.Count)
+                {
+                    Vector256<T> targetVector = Vector256.Create(value);
+                    ref T oneVectorAwayFromEndMinus1 = ref Unsafe.Subtract(ref end, Vector256<T>.Count - 1);
+                    do
+                    {
+                        count += BitOperations.PopCount(Vector256.Equals(Vector256.LoadUnsafe(ref current), targetVector).ExtractMostSignificantBits());
+                        current = ref Unsafe.Add(ref current, Vector256<T>.Count);
+                    }
+                    while (Unsafe.IsAddressLessThan(ref current, ref oneVectorAwayFromEndMinus1));
+
+                    if (Unsafe.IsAddressLessThan(ref current, ref Unsafe.Subtract(ref end, Vector128<T>.Count - 1)))
```

Comment:
@tannergooding given this pattern where the Vector256, Vector128, and presumably Vector512 implementations are otherwise identical -- do we have a pattern planned that would allow them to  be collapsed? this seems like a place where you want to just use Vector, but still get the benefits of a fixed size vector, without writing all this out.

Reply:
We've talked about a few approaches:
- Allowing `Vector<T>` to play nicely with dynamic PGO
- Having an `IVector<T>` that would let you write the code once and call it three times with the three different vector sizes
- For specific kinds of situations, e.g. IndexOf, experiment with using a pattern based on an interface that lets you express just the kernel of the search for each of the vector sizes and have the calling API supply all the boilerplate around it.

Reply:
@danmoseley, yes. I need to take my findings from https://github.com/dotnet/runtime/pull/76423 and create an API proposal.

I have most of one complete, but not all of one yet